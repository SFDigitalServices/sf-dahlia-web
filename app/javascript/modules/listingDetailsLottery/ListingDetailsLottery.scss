.modal {
  @apply max-w-0;
}

.modal__inner {
  @apply p-0;
}

.modal__close {
  z-index: 100;
}
