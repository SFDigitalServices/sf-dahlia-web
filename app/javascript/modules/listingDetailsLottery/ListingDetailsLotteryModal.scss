.lottery-modal {
    .lottery-results-form .input {
        border-radius: initial
    }
}