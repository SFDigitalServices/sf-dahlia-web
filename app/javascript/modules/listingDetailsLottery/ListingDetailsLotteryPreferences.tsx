import React from "react"
import { t } from "@bloom-housing/ui-components"
import dayjs from "dayjs"
import { RailsLotteryBucketsDetails } from "../../api/types/rails/listings/RailsLotteryBucketsDetails"
import { RailsListing } from "../listings/SharedHelpers"

export interface ListingDetailsLotteryPreferencesProps {
  listing: RailsListing
  lotteryBucketsDetails: RailsLotteryBucketsDetails
}

export const ListingDetailsLotteryPreferences = ({
  listing,
  lotteryBucketsDetails,
}: ListingDetailsLotteryPreferencesProps) => {
  return (
    <div className="text-tiny">
      <p className="py-4 text-sm">{t("lottery.buckets_intro")}</p>
      <h2 className="text-caps-underline">{t("lottery.housing_preferences")}</h2>
      <p className="border-b border-gray-450 mb-4 pb-4 text-gray-700 text-sm">
        {t("lottery.ranking_order_note")}
      </p>
      {lotteryBucketsDetails.lotteryBuckets
        .filter((bucket) => bucket.preferenceName !== "generalLottery")
        .map((bucket, index, arr) => (
          <>
            <h3
              className="font-sans font-semibold text-tiny tracking-wide uppercase"
              key={bucket.preferenceOrder}
            >
              {bucket.preferenceName}
            </h3>
            <p className="text-tiny">
              {t("lottery.up_to_x_units_available", { units: bucket.unitsAvailable })}
            </p>
            <p className="text-gray-700 text-tiny">
              {t("lottery.number_applicants_qualified_for_preference", {
                number: bucket.totalSubmittedApps,
              })}
            </p>
            <hr className={arr.length - 1 === index ? "mt-4" : "my-4"} />
          </>
        ))}
      <div className="bg-gray-100 border-b mb-4 p-4">
        <hr className="border-b-4 border-primary" />
      </div>
      {lotteryBucketsDetails.lotteryBuckets
        .filter((bucket) => bucket.preferenceName === "generalLottery")
        .map((bucket) => (
          <div key={bucket.preferenceOrder}>
            <h3 className="font-sans font-semibold text-tiny tracking-wide uppercase">
              {t("lottery.general_pool")}
            </h3>
            <p className="text-tiny">{t("lottery.any_remaining_units")}</p>
            <p className=" border-b pb-2 text-gray-700 text-tiny">
              {t("lottery.number_applicants_qualified_for_general_pool", {
                number: bucket.totalSubmittedApps,
              })}
            </p>
          </div>
        ))}
      <div className="bg-gray-100 mb-2 py-4 text-center">
        <a href="http://sfmohcd.org/housing-preference-programs" target="_blank">
          {t("lottery.read_about_preferences")}
        </a>
      </div>
      <div className="text-center">
        <p className="py-2">
          {t("lottery.results_from", {
            date: dayjs(listing.Lottery_Results_Date).format("MMM D, YYYY"),
          })}
        </p>
        <p className="py-2">
          <a href={listing.LotteryResultsURL} target="_blank">
            {t("lottery.lottery_results_pdf_link")}
          </a>
        </p>
        <p className="text-gray-700 py-2">
          {t("lottery.unsorted_results_note", {
            date: dayjs(listing.Lottery_Date).format("MMM D, YYYY"),
          })}
        </p>
      </div>
    </div>
  )
}
