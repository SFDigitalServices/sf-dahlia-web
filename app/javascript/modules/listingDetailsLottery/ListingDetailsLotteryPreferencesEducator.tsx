import React from "react"
import { Heading, t } from "@bloom-housing/ui-components"
import type { RailsLotteryResult } from "../../api/types/rails/listings/RailsLotteryResult"
import { defaultIfNotTranslated, renderMarkup } from "../../util/languageUtil"
import { preferenceNameHasVeteran } from "../../util/listingUtil"
import { LOTTERY_RANKING_VIDEO_URL } from "../constants"
import { RailsLotteryBucket } from "../../api/types/rails/listings/RailsLotteryBucket"

export interface ListingDetailsLotteryPreferencesProps {
  lotteryBucketsDetails: RailsLotteryResult
  isEducatorOne: boolean
}
interface PreferencesProps {
  unitsAvailable: number
  totalSubmittedApps: number
  preferenceName: string
  numVeteranApps: number
  lastPref: boolean
}

const LotteryPreferences = ({
  unitsAvailable,
  totalSubmittedApps,
  preferenceName,
  numVeteranApps,
  lastPref,
}: PreferencesProps) => (
  <div>
    <div className="px-8">
      <Heading className="font-sans font-semibold text-xs tracking-wide uppercase" priority={3}>
        {defaultIfNotTranslated(
          `listings.lotteryPreference.${preferenceName}.title`,
          preferenceName
        )}
      </Heading>
      <p className="text-sm">{t("lottery.upToXUnitsAvailable", { units: unitsAvailable })}</p>
      {numVeteranApps > 0 ? (
        <>
          <p className="text-gray-700 text-sm">
            {t("lottery.numberApplicantsQualifiedForPreference.veteran", {
              number: numVeteranApps,
              plural: numVeteranApps > 1 ? "s" : "",
            })}
          </p>
          <p className="text-gray-700 text-sm">
            {t("lottery.numberApplicantsQualifiedForPreference.nonVeteran", {
              number: totalSubmittedApps,
              plural: totalSubmittedApps > 1 ? "s" : "",
            })}
          </p>
        </>
      ) : (
        <p className="text-gray-700 text-sm">
          {t("lottery.numberApplicantsQualifiedForPreference", {
            number: totalSubmittedApps,
            plural: totalSubmittedApps > 1 ? "s" : "",
          })}
        </p>
      )}
    </div>
    <hr className={lastPref ? "mt-4" : "my-4"} />
  </div>
)

const LotteryBucketHeader = ({ headerKey }: { headerKey: string }) => {
  return (
    <header className="border-b border-gray-450 mb-4">
      <Heading styleType="underlineWeighted" className="mx-8 text-xs" priority={2}>
        {t(headerKey)}
      </Heading>
    </header>
  )
}

const LotteryBucketsContent = ({
  buckets,
  veteranAppsByBucketMap,
  preferenceNameKey,
  lastPref,
}: {
  buckets: Array<RailsLotteryBucket>
  veteranAppsByBucketMap: object
  preferenceNameKey?: string
  lastPref?: boolean
}) => {
  return buckets.map((bucket, index, arr) => {
    const preferenceName = preferenceNameKey
      ? bucket.preferenceName.replace(preferenceNameKey, "")
      : bucket.preferenceName

    // We want to display the number of veteran applications under the associated preference bucket
    // so we will look up the veteran applications by the veteran preference short code that we derive
    // from the preference short code by adding in the 'V-'
    // e.g. The number of veteran applications for 'T1-COP' are stored in veteranAppsByBucketMap
    // under the key 'T1-V-COP', so we insert 'V-' using replace to get the veteran preference short code
    const veteransPreferenceShortCode = bucket.preferenceShortCode
    const numVeteranApps =
      veteranAppsByBucketMap[veteransPreferenceShortCode.replace("-", "-V-")] ?? 0

    return (
      <LotteryPreferences
        unitsAvailable={bucket.unitsAvailable}
        totalSubmittedApps={bucket.totalSubmittedApps}
        preferenceName={preferenceName}
        numVeteranApps={numVeteranApps}
        lastPref={lastPref && arr.length - 1 === index}
      />
    )
  })
}

export const ListingDetailsLotteryPreferencesEducator = ({
  lotteryBucketsDetails,
  isEducatorOne,
}: ListingDetailsLotteryPreferencesProps) => {
  // First, map all the veteran buckets to their corresponding educator bucket
  const veteranAppsByBucketMap = {}
  lotteryBucketsDetails.lotteryBuckets
    .filter((lotteryBucket) => preferenceNameHasVeteran(lotteryBucket.preferenceName))
    .forEach((lotteryBucket) => {
      // stores the number of veteran applications under the preference short code
      // e.g. The code for the veteran tier 1 cop bucket is 'T1-V-COP' - when we
      // later query for the veteran applications associated with the 'T1-COP' bucket
      // we will insert 'V-' to 'T1-COP' to get the lookup key
      veteranAppsByBucketMap[lotteryBucket.preferenceShortCode] = lotteryBucket.totalSubmittedApps
    })

  console.log(veteranAppsByBucketMap)
  // Then, get all the tier 1 educator buckets, excluding the tier 1 veteran buckets
  const tier1EducatorBucketWithoutVeteran = lotteryBucketsDetails.lotteryBuckets.filter(
    (bucket) =>
      bucket.preferenceShortCode &&
      bucket.preferenceShortCode.includes("T1") &&
      !bucket.preferenceShortCode.includes("V") &&
      bucket.totalSubmittedApps
  )

  // Then, get all the tier 2 educator buckets, excluding the tier 1 veteran buckets
  const tier2EducatorBucketWithoutVeteran = lotteryBucketsDetails.lotteryBuckets.filter(
    (bucket) =>
      bucket.preferenceShortCode &&
      bucket.preferenceShortCode.includes("T2") &&
      !bucket.preferenceShortCode.includes("V") &&
      bucket.totalSubmittedApps
  )

  // Then, get all the other non-educator buckets, excluding the veteran buckets
  // This only applies for SCV 2 & 3 (SCV 1 only allows educators to apply)
  const nonEducatorsWithoutVeterans = lotteryBucketsDetails.lotteryBuckets.filter(
    (bucket) =>
      bucket.preferenceShortCode &&
      !bucket.preferenceShortCode.includes("T1") &&
      !bucket.preferenceShortCode.includes("T2") &&
      !bucket.preferenceShortCode.includes("V") &&
      bucket.totalSubmittedApps
  )

  // Then, get the general lottery bucket
  // This only applies for SCV 2 & 3 (SCV 1 only allows educators to apply)
  const generalLottery = lotteryBucketsDetails.lotteryBuckets.filter(
    (bucket) => bucket.preferenceName === "generalLottery"
  )

  return (
    <div className="text-sm">
      <header className="border-b border-gray-450 mb-4">
        <Heading styleType="underlineWeighted" className="mx-8" priority={2}>
          {t("lottery.housingPreferences.educator")}
        </Heading>
        <div className="pb-4 text-gray-700 text-xs mx-8 translate">
          {renderMarkup(
            `${t("lottery.rankingOrderNote", {
              lotteryRankingVideoUrl: LOTTERY_RANKING_VIDEO_URL,
            })}`
          )}
        </div>
        <div className="pb-4 text-gray-700 text-xs mx-8">
          {t("lottery.rankingOrderNote.veterans")}
        </div>
      </header>
      <LotteryBucketHeader headerKey="listings.lotteryPreference.sfusd.educators.title" />
      <LotteryBucketsContent
        buckets={tier1EducatorBucketWithoutVeteran}
        veteranAppsByBucketMap={veteranAppsByBucketMap}
        preferenceNameKey="Tier 1 "
      />
      <LotteryBucketHeader headerKey="listings.lotteryPreference.sfusd.allOtherEmployees.title" />
      <LotteryBucketsContent
        buckets={tier2EducatorBucketWithoutVeteran}
        veteranAppsByBucketMap={veteranAppsByBucketMap}
        preferenceNameKey="Tier 2 "
        lastPref={isEducatorOne}
      />
      {!isEducatorOne && (
        <>
          <LotteryBucketHeader headerKey="lottery.generalPublic" />
          <LotteryBucketsContent
            buckets={nonEducatorsWithoutVeterans}
            veteranAppsByBucketMap={veteranAppsByBucketMap}
            lastPref
          />
          <div className="bg-gray-100 border-b mb-4 p-4">
            <hr className="border-b-4 border-primary" />
          </div>
          {generalLottery.map((bucket) => (
            <React.Fragment key={bucket.preferenceOrder}>
              <div className="px-8">
                <Heading
                  className="font-sans font-semibold text-sm tracking-wide uppercase"
                  priority={3}
                >
                  {t("lottery.generalPool")}
                </Heading>
                <p className="mb-1 text-sm">{t("lottery.anyRemainingUnits")}</p>
                <p className="text-gray-700 text-sm">
                  {t("lottery.numberApplicantsQualifiedForGeneralPool", {
                    number: bucket.totalSubmittedApps,
                  })}
                </p>
              </div>
              <hr className="mt-4" />
            </React.Fragment>
          ))}
        </>
      )}
    </div>
  )
}
