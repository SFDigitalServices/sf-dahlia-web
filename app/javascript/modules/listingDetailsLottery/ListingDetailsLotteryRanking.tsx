import React from "react"
import { ExpandableContent, Heading, Icon, t } from "@bloom-housing/ui-components"
import { PREFERENCES } from "../constants"
import { ListingDetailsLotteryResultsRow } from "./ListingDetailsLotteryResultsRow"
import Link from "../../navigation/Link"
import type { RailsLotteryResult } from "../../api/types/rails/listings/RailsLotteryResult"
import { getSfGovUrl, renderMarkup } from "../../util/languageUtil"

interface ListingDetailsLotteryRankingProps {
  lotteryResult: RailsLotteryResult
  listingIsEducator: boolean
}

interface TooltipProps {
  text: string
}

const Tooltip = ({ text }: TooltipProps) => {
  return (
    <div className="bg-blue-100 border-b border-gray-450 flex mb-4 py-4">
      <Icon className="mx-4" size="medium" symbol="info" />
      <p>{text}</p>
    </div>
  )
}
export const ListingDetailsLotteryRanking = ({
  lotteryResult,
  listingIsEducator,
}: ListingDetailsLotteryRankingProps) => {
  const preferenceBuckets = lotteryResult?.lotteryBuckets.filter((bucket) => {
    if (!bucket.preferenceResults[0]) {
      return false
    }
    return bucket.preferenceName !== "generalLottery" && bucket.preferenceResults[0].preferenceRank
  })
  const applicantSelectedForPreference = preferenceBuckets?.length > 0
  const applicantHasCertOfPreference = preferenceBuckets.some(
    (bucket) => bucket.preferenceName === PREFERENCES.certificateOfPreference
  )
  const generalLotteryBucket = lotteryResult?.lotteryBuckets.find(
    (bucket) => bucket.preferenceName === "generalLottery"
  )

  return (
    lotteryResult && (
      <div className="lottery-ranking text-sm">
        <header className="px-8">
          <Heading styleType="underlineWeighted" priority={2}>
            {applicantSelectedForPreference
              ? t("lottery.rankingTitle")
              : t("lottery.rankingTitle.noPreference")}
          </Heading>
          <div className="border-b border-gray-450 pb-4 text-gray-700 translate">
            {renderMarkup(
              `${t("lottery.rankingOrderNote", {
                lotteryRankingVideoUrl: "https://www.youtube.com/watch?v=4ZB35gagUl8",
              })}`
            )}
          </div>
        </header>

        {applicantSelectedForPreference && !applicantHasCertOfPreference && (
          <Tooltip text={t("lottery.rankingPreferenceConsiderationNote")} />
        )}
        {!applicantSelectedForPreference && (
          <Tooltip text={t("lottery.rankingPreferencesConsideredOverGeneralNote")} />
        )}
        {preferenceBuckets?.map((bucket) => (
          <ListingDetailsLotteryResultsRow
            bucket={bucket}
            key={bucket.preferenceName}
            listingIsEducator={listingIsEducator}
          />
        ))}
        {!applicantSelectedForPreference && (
          <ListingDetailsLotteryResultsRow
            bucket={generalLotteryBucket}
            listingIsEducator={listingIsEducator}
          />
        )}
        <div className="px-8">
          <ExpandableContent
            className="pb-4"
            strings={{
              readMore: t("lottery.nextStepsTitle"),
              readLess: t("lottery.nextStepsTitle"),
            }}
          >
            <div className="text-gray-700">
              <p className="mb-2">
                <Link
                  className="text-blue-700"
                  external={true}
                  href={getSfGovUrl("https://sf.gov/after-rental-housing-lottery", 12704)}
                  target="_blank"
                >
                  {t("lottery.nextStepsLearnMore")}
                </Link>
              </p>
              <p className="mb-2">{t("lottery.nextStepsP1")}</p>
              <p className="mb-2">{t("lottery.nextStepsP2")}</p>
              <p className="mb-2">{t("lottery.nextStepsP3")}</p>
            </div>
          </ExpandableContent>
        </div>
      </div>
    )
  )
}
