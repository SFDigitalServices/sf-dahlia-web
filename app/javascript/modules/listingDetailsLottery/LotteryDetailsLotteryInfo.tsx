import React from "react"
import { SidebarBlock, t } from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import { localizedFormat, renderWithInnerHTML } from "../../util/languageUtil"
import { isLotteryComplete, isOpen } from "../../util/listingUtil"

export interface ListingDetailsLotteryInfoProps {
  listing: RailsListing
}

export const ListingDetailsLotteryInfo = ({ listing }: ListingDetailsLotteryInfoProps) => {
  if (isOpen(listing) || !listing.Lottery_Date || isLotteryComplete(listing)) {
    return null
  }

  return (
    <>
      <SidebarBlock title={t("label.lottery")}>
        <p className="flex justify-between mb-4">
          <span>{localizedFormat(listing.Lottery_Date, "LL")}</span>
          <span className="font-bold">{localizedFormat(listing.Lottery_Date, "LT")}</span>
        </p>
        {listing.Lottery_Venue && renderWithInnerHTML(listing.Lottery_Venue)}
        {process.env.COVID_UPDATE !== "true" && <p className="mt-4">{t("label.preLotteryInfo")}</p>}
      </SidebarBlock>

      <SidebarBlock title={t("lottery.lotteryResults")}>
        <p className="mb-4">{localizedFormat(listing.Lottery_Results_Date, "LL")}</p>
        <p>{t("lottery.completeResultsWillBePosted")}</p>
      </SidebarBlock>
    </>
  )
}
