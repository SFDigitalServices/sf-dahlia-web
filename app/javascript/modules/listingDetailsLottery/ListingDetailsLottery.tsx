import React, { useEffect, useState } from "react"
import {
  AppearanceSizeType,
  AppearanceStyleType,
  Button,
  Heading,
  Modal,
  t,
} from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import { isLotteryComplete } from "../../util/listingUtil"
import { getLotteryBucketDetails } from "../../api/listingApiService"
import { RailsLotteryResult } from "../../api/types/rails/listings/RailsLotteryResult"
import { ListingDetailsLotteryModal } from "./ListingDetailsLotteryModal"
import { localizedFormat, renderInlineWithInnerHTML } from "../../util/languageUtil"

export interface ListingDetailsLotteryProps {
  listing: RailsListing
}

export const ListingDetailsLottery = ({ listing }: ListingDetailsLotteryProps) => {
  const [lotteryBucketDetails, setLotteryBucketDetails] = useState<RailsLotteryResult>()
  const [isModalOpen, setIsModalOpen] = useState(false)

  useEffect(() => {
    void getLotteryBucketDetails(listing.Id).then((lotteryBucketDetails) => {
      setLotteryBucketDetails(lotteryBucketDetails)
    })
  }, [listing.Id])

  return (
    isLotteryComplete(listing) && (
      <div className="border-b pt-4 text-center">
        <Heading className="mb-4" priority={4}>
          {t("lottery.lotteryResults")}
        </Heading>
        <p className="mb-4 text-sm uppercase">
          {localizedFormat(listing.Lottery_Results_Date, "LL")}
        </p>
        <div className="bg-gray-100 py-4">
          {listing.Lottery_Summary && (
            <div className="mb-3 mx-2 text-gray-700 text-sm">
              {renderInlineWithInnerHTML(listing.Lottery_Summary)}
            </div>
          )}
          <Button
            size={AppearanceSizeType.small}
            styleType={AppearanceStyleType.primary}
            onClick={() => {
              setIsModalOpen(true)
            }}
          >
            {t("listings.lottery.viewLotteryResults")}
          </Button>
        </div>

        {lotteryBucketDetails && (
          <Modal onClose={() => setIsModalOpen(false)} open={isModalOpen} title="">
            <ListingDetailsLotteryModal
              listing={listing}
              lotteryBucketDetails={lotteryBucketDetails}
            />
          </Modal>
        )}
      </div>
    )
  )
}
