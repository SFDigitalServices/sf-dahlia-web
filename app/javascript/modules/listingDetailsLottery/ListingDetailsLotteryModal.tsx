import React, { useState } from "react"
import { Field, Heading, Icon, IconFillColors, Modal, t } from "@bloom-housing/ui-components"
import { useForm } from "react-hook-form"
import { RailsLotteryBucketsDetails } from "../../api/types/rails/listings/RailsLotteryBucketsDetails"
import { RailsListing } from "../listings/SharedHelpers"
import { ListingDetailsLotteryPreferences } from "./ListingDetailsLotteryPreferences"
import "./ListingDetailsLotteryModal.scss"
import { getLotteryResults } from "../../api/listingApiService"
import { RailsLotteryRanking } from "../../api/types/rails/listings/RailsLotteryRanking"
import { ListingDetailsLotteryRanking } from "./ListingDetailsLotteryRanking"
import { ListingDetailsLotteryModalFooter } from "./ListingDetailsLotteryModalFooter"

export enum STATES {
  INITIAL_STATE,
  LOADING,
  API_ERROR,
  INVALID_LOTTERY_NUMBER,
  SEARCH_RESULT_FOUND,
}

interface ListingDetailsLotteryModalProps {
  isOpen: boolean
  listing: RailsListing
  lotteryBucketDetails: RailsLotteryBucketsDetails
  onClose: () => void
}

export const ListingDetailsLotteryModal = ({
  isOpen,
  listing,
  lotteryBucketDetails,
  onClose,
}: ListingDetailsLotteryModalProps) => {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { errors, handleSubmit, register } = useForm({ reValidateMode: "onSubmit" })
  const lotteryNumberField = "lotterySearchNumber"
  const [lotterySearchResult, setLotterySearchResult] = useState<RailsLotteryRanking>()

  const [lotteryModalStatus, setLotteryModalStatus] = useState<STATES>(STATES.INITIAL_STATE)

  const onModalClose = () => {
    setLotteryModalStatus(STATES.INITIAL_STATE)
    onClose()
  }

  const onSubmit = (data: { lotterySearchNumber: string }) => {
    const { lotterySearchNumber } = data
    setLotteryModalStatus(STATES.LOADING)

    void getLotteryResults(listing.Id, lotterySearchNumber).then((lotterySearchResults) => {
      if (!lotterySearchResults) {
        // if we didn't get a result, general search api error
        setLotteryModalStatus(STATES.API_ERROR)
        return
      }

      const { lotteryBuckets } = lotterySearchResults
      if (
        // if we got a result but there are no buckets or none of the buckets have pref
        // results, then we don't have a valid lottery number
        !lotteryBuckets ||
        lotteryBuckets.filter((bucket) => bucket.preferenceResults.length > 0).length === 0
      ) {
        setLotteryModalStatus(STATES.INVALID_LOTTERY_NUMBER)
      } else {
        setLotteryModalStatus(STATES.SEARCH_RESULT_FOUND)
        setLotterySearchResult(lotterySearchResults)
      }
    })
  }

  let content: JSX.Element
  switch (lotteryModalStatus) {
    case STATES.LOADING: {
      content = (
        <div className="text-center">
          <Icon size="large" symbol="spinner" />
        </div>
      )
      break
    }
    case STATES.INVALID_LOTTERY_NUMBER: {
      content = (
        <div className="my-6 text-gray-700">
          <p className="mb-2">{t("lottery.lotteryNumberNotFoundP2")}</p>
          <p className="mb-2">{t("lottery.lotteryNumberNotFoundP3")}</p>
        </div>
      )
      break
    }
    case STATES.SEARCH_RESULT_FOUND: {
      content = lotterySearchResult && (
        <ListingDetailsLotteryRanking lotteryRanking={lotterySearchResult} />
      )
      break
    }
    case STATES.INITIAL_STATE: {
      content = <ListingDetailsLotteryPreferences lotteryBucketsDetails={lotteryBucketDetails} />
      break
    }
  }

  const errorMessage =
    (errors[lotteryNumberField]?.type === "required" && t("lottery.lotteryNumberNotValid")) ||
    (lotteryModalStatus === STATES.INVALID_LOTTERY_NUMBER &&
      t("lottery.lotteryNumberNotFoundP1")) ||
    (lotteryModalStatus === STATES.API_ERROR && t("error.lotteryRankingSearch"))

  return (
    <Modal onClose={onModalClose} open={isOpen} title="">
      <div className="lottery-modal w-96">
        <header className="pb-4 text-center">
          <Heading>{t("lottery.lotteryResults")}</Heading>
          <h2 className="font-sans font-semibold text-sm uppercase">{listing.Name}</h2>
        </header>
        <form
          className="bg-gray-100 flex lottery-results-form mb-4 px-6 py-4"
          onSubmit={handleSubmit(onSubmit)}
        >
          <Field
            error={
              errors[lotteryNumberField] ||
              lotteryModalStatus === STATES.API_ERROR ||
              lotteryModalStatus === STATES.INVALID_LOTTERY_NUMBER
            }
            className="w-full"
            errorMessage={
              errorMessage ||
              (errors[lotteryNumberField]?.type === "required" &&
                t("lottery.lotteryNumberNotValid"))
            }
            register={register}
            name={lotteryNumberField}
            placeholder={t("lottery.enterLotteryNumber")}
            type="text"
            validation={{ required: true }}
          />
          <button className="bg-blue-600 h-12 mt-2 p-3">
            <Icon fill={IconFillColors.white} size="medium" symbol="right" />
          </button>
        </form>
        {content}
        <ListingDetailsLotteryModalFooter
          lotteryModalStatus={lotteryModalStatus}
          listing={listing}
        />
      </div>
    </Modal>
  )
}
