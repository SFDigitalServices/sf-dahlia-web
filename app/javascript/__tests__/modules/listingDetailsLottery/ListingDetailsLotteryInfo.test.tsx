import React from "react"
import { render } from "@testing-library/react"
import { openSaleListing } from "../../data/RailsSaleListing/listing-sale-open"
import { ListingDetailsLotteryInfo } from "../../../modules/listingDetailsLottery/LotteryDetailsLotteryInfo"
import { lotteryCompleteRentalListing } from "../../data/RailsRentalListing/listing-rental-lottery-complete"
import { closedRentalListing } from "../../data/RailsRentalListing/listing-rental-closed"

describe("ListingDetailsLotteryInfo", () => {
  it("does not display when listing has no lottery date", () => {
    // mock a listing with no lottery date
    const testListing = { ...openSaleListing, Lottery_Date: null }

    const { asFragment } = render(<ListingDetailsLotteryInfo listing={testListing} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it("does not display when listing is open", () => {
    const { asFragment } = render(<ListingDetailsLotteryInfo listing={openSaleListing} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it("displays when listing has upcoming lottery", () => {
    const testListing = { ...closedRentalListing }

    const { asFragment } = render(<ListingDetailsLotteryInfo listing={testListing} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it("does not display when listing has a completed lottery", () => {
    const { asFragment } = render(
      <ListingDetailsLotteryInfo listing={lotteryCompleteRentalListing} />
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
