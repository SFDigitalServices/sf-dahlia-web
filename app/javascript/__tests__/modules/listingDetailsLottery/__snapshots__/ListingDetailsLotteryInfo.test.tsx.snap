// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingDetailsLotteryInfo displays when listing has upcoming lottery 1`] = `
<DocumentFragment>
  <div
    class="border-b border-gray-400 md:border-b-0"
  >
    <section
      class="aside-block "
    >
      <h4
        class="text__underline-weighted"
      >
        Lottery
      </h4>
      <div
        class="text-sm text-gray-750"
      >
        <p
          class="flex justify-between mb-4"
        >
          <span>
            March 22, 2023
          </span>
          <span
            class="font-bold"
          >
            6:00 pm
          </span>
        </p>
        <div
          class="text-gray-700"
        >
          <div>
            <p>
              The lottery will not be in person, but instead will be a virtual lottery.
            </p>
            <p>
              Â 
            </p>
            <p>
              Come back to this page on the day of the lottery for a link you can click on to watch the lottery live.
            </p>
            <p>
              <br />
            </p>
            <p>
              <br />
            </p>
          </div>
        </div>
        <p
          class="mt-4 text-gray-700"
        >
          An open lottery will be held to draw numbers to determine selection. Attendance is optional.
        </p>
      </div>
    </section>
  </div>
  <div
    class="border-b border-gray-400 md:border-b-0"
  >
    <section
      class="aside-block "
    >
      <h4
        class="text__underline-weighted"
      >
        Lottery Results
      </h4>
      <div
        class="text-sm text-gray-750"
      >
        <p
          class="mb-4"
        >
          March 23, 2023
        </p>
        <p
          class="text-gray-700"
        >
          Complete lottery results will be posted on this date by 5 pm.
        </p>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`ListingDetailsLotteryInfo does not display when listing has a completed lottery 1`] = `<DocumentFragment />`;

exports[`ListingDetailsLotteryInfo does not display when listing has no lottery date 1`] = `<DocumentFragment />`;

exports[`ListingDetailsLotteryInfo does not display when listing is open 1`] = `<DocumentFragment />`;
